src/NServiceBus.Core/Endpoint.cs

```diff
@@ -1,21 +1,59 @@
 namespace NServiceBus
 {
     using System.Threading.Tasks;
+    using ObjectBuilder;
 
     /// <summary>
     /// Provides factory methods for creating and starting endpoint instances.
     /// </summary>
     public static class Endpoint
     {
+        /// <summary>
+        /// Prepares an endpoint to run with and externally managed container.
+        /// </summary>
+        /// <param name="configuration">The endpoint configuration.</param>
+        /// <param name="configureComponents">The registration api adapter for the external container.</param>
+        /// <returns>The prepared endpoint.</returns>
+        public static PreparedEndpoint Prepare(EndpointConfiguration configuration, IConfigureComponents configureComponents)
```

Comment:
Not sure if this changes anything but I envision that this will replace
.Create in the long run?

Reply:
Why a new method then? Why can't this be in Create instead?

Reply:
You mean to make this a en overload of the existing .Create?

Reply:
Since we need to keep the old create for backwards compatibility

Reply:
My concern is introducing yet another API to configure and start things, meaning we have to document it and answer customer questions about it. Surely there's a simpler way to achieve this without needing to add more configuration API surface area.

`Prepare` as the word still sounds off. If it were replacing `Create`, sure, that sounds better. But if it's replacing `Create` why can't the logic inside `Prepare` simply be done inside the `Create`?


```
public static Task<IStartableEndpoint> Create(EndpointConfiguration configuration)
{
    if (externallyManagedContainerHasBeenSpecified)
    {
        // do the stuff here
    }

    // other create stuff here
}
```

I'm not sure what backwards compatibility stuff would happen that can't be done behind a feature toggle.

Reply:
One of the requirements is to not require `async` at this stage so we can't use the existing method.

We also need the user to pass us the builder so it felt more discoverabe to make it a parameter instead of adding some kind of `EndpointConfiguration.RegisterBuilder()` that we would have to check that it was called at runtime?

The option I see is another overload:

```
public static Task<IStartableEndpoint> Create(EndpointConfiguration configuration)

public static PreparedEndpoint Create(EndpointConfiguration configuration, IConfigureComponents configureComponents)
```

But I have a bad feeling about mixing async and non async like that?

Reply:
I think as long as we want to push changes without unnecessarily having to bump the core major we are faced with difficult decisions like solving one pain point but potentially introducing another source of confusion due to having multiple APIs that can lead to the same result. Like @andreasohlund mentioned having a non-async method is crucial for the registration APIs. The current API approach makes it clear which API has to be called at what stage. I'm with andreas as well that introducing a create method that has a different return type as well as async based on the provided parameter would be even worse and we would seriously have to consider using the `Async` suffix again.

One thing that we've discussed in the past when we design the v6 APIs around `Endpoint` was that those APIs tend to be used by a different developer type. All the APIs around endpoint configuration require more understanding of plumping which is naturally something more lead devs do than "regular developers" (sorry for that term).

Reply:
> those APIs tend to be used by a different developer type

For some types of customers, sure. Not for the types of customers that raise support cases about this kind of stuff though.

>  it felt more discoverabe to make it a parameter instead of adding some kind of `EndpointConfiguration.RegisterBuilder()`

I'm not sure I agree. Every other dial that we allow users to configure is via an EndpointConfiguration method. What makes it better to change this method here and introduce a new way of starting endpoints for this?

If a customer has an existing piece of code using `Create` and `Start`, instead of adding a `.RegisterBuilder()` before that, they will now have to change how they start endpoints and use `Prepare` and a new `PreparedEndpoint` type. It could be quite a painful change depending on how they've done their configure + start code.

Reply:
The issue with async is still a blocker for me, Since we can't be async at the "prepare" stage I believe we need a new method name to not mix async and sync on "the same method".

Regarding using a parameter or a  `.RegisterBuilder()` => seems like we're mostly talking about personal preference at this stage but my personal opinion is that I rather add an API that users can't screw up now rather than following "what we've done before" since I'm pretty sure that wasn't a conscious decision. What do you others think?

Reply:
Also note that DevEx will ship an integration package for MS DI in the next phase of all this so users are unlikely to be exposed to this API directly going forwards

Reply:
> seems like we're mostly talking about personal preference

💯. So let's be very clear here: "what we've done before" is everywhere - conscious decision or not. And changing the paradigm away from "what we've done before" should not just be done based on personal preference. It has an impact on our customers. It has an impact on how much time they have to spend on enabling this change in their systems. It has an impact on the cognitive load for us when it comes to support cases. A change away from "what we've done before" should be a measured, conscious, planned change that should also have other configuration options planned to make similar changes too.

With semver rules, changing away from "what we've done before" is a change that could quite happily slide in under the same major. I am arguing that changing a paradigm should most definitely not be done in a minor though due to the change in docs and mental model involved. In other words, having a different way of doing things is a major change. The fact that it slides in as "binary compatible and can be done in a minor" is hiding the impact of a move away from "what we've done before" which happens to be everywhere.

Ultimately I'm just a bystander giving feedback, but I'm not seeing any arguments that have convinced me that a change like this in a minor is for the better.

Reply:
> so users are unlikely to be exposed to this API directly going forwards

Even more reason to not break things now. Why make customers change to prepare and preparedendpoints now when they won't have to in the future?

Reply:
> Not for the types of customers that raise support cases about this kind of stuff though.

Can you explain a bit more what type of customers you've been seeing struggle? 

If I understand you correctly then your strongest argument is that because previously the container was part of the endpoint configuration API it is most likely that customers will have the container already available there thus logically it should be passed somehow into the endpoint configuration?

And you are also saying that it is much more likely that endpoint configuration phase is distinct in the code from the `Endpoint.Create` and `Start` usage thus it would make it painful for them to float the container to that part of the code?

Reply:
> Can you explain a bit more what type of customers you've been seeing struggle?

Sure, in your words:

> "regular developers" (sorry for that term)

Your separation of who uses what bits of code is not based on anything. We can gladly walk through a bunch of support cases and chats. I'd wager that taking a few at random will show a non-zero percentage of "regular developers" getting lost in configuration code.

Reply:
I've scheduled a call to talk this through. Will update here with the outcome

Reply:
Syncing with @danielmarbach @WilliamBZA and later @SzymonPobiega we came up with a tweak that is more aligned with what we do today and will also solve @timbussmann concerns as it seems.

@SzymonPobiega will make the changes so stay tuned